src/fsharp/FSharp.Compiler.Unittests/HashIfExpression.fs

```diff
@@ -0,0 +1,268 @@
+                "!ONE"                      , !!ONE
+                "!!ONE"                     , !! (!!ONE)
+                "DEBUG"                     , (IfdefId "DEBUG")
+                "!DEBUG"                    , !! (IfdefId "DEBUG")
+                "O_s1"                      , IfdefId "O_s1"
+                "(ONE)"                     , (ONE)
+                "ONE&&TWO"                  , ONE &&& TWO
+                "ONE||TWO"                  , ONE ||| TWO
+                "( ONE && TWO )"            , ONE &&& TWO
+                "ONE && TWO && THREE"       , (ONE &&& TWO) &&& THREE
+                "ONE || TWO || THREE"       , (ONE ||| TWO) ||| THREE
+                "ONE || TWO && THREE"       , ONE ||| (TWO &&& THREE)
+                "ONE && TWO || THREE"       , (ONE &&& TWO) ||| THREE
+                "ONE || (TWO && THREE)"     , ONE ||| (TWO &&& THREE)
+                "ONE && (TWO || THREE)"     , ONE &&& (TWO ||| THREE)
+                "!ONE || TWO && THREE"      , (!!ONE) ||| (TWO &&& THREE)
+                "ONE && !TWO || THREE"      , (ONE &&& (!!TWO)) ||| THREE
+                "ONE || !(TWO && THREE)"    , ONE ||| (!!(TWO &&& THREE))
+                "true"                      , IfdefId "true"
+                "false"                     , IfdefId "false"
+            |]
+
+        let failures    = ResizeArray<string> ()
+        let fail        = failures.Add
+
+        for test,expected in positiveTestCases do
+            for prelude in preludes do
+                let test = prelude + test
+                for epilogue in epilogues do
+                    let test = test + epilogue
+                    try
+                        let expr    = parser test
+
+                        if expected <> expr then
+                            fail <| sprintf "'%s', expected %A, actual %A" test (exprAsString expected) (exprAsString expr)
+                    with
+                    | e ->  fail <| sprintf "'%s', expected %A, actual %s,%A" test (exprAsString expected) (e.GetType().Name) e.Message
+
+
+        let fs =
+            failures
+            |> Seq.append (errors   |> Seq.map (fun pe -> pe.DebugDisplay ()))
+            |> Seq.append (warnings |> Seq.map (fun pe -> pe.DebugDisplay ()))
+            |> Seq.toArray
+
+        let failure = String.Join ("\n", fs)
+
+        Assert.AreEqual("", failure)
+
+        ()
+
+    [<Test>]
+    member this.NegativeParserTestCases()=
+
+        let errors, warnings, parser = createParser ()
+
+        let negativeTests =
+            [|
+                ""
+                "!"
+                "&&"
+                "||"
+                "@"
+                "ONE ONE"
+                "ONE@"
+                "@ONE"
+                "$"
+                "ONE$"
+                "$ONE"
+                "ONE!"
+                "(ONE"
+                "ONE)"
+                // TODO: Investigate why this raises a parse failure
+                // "(ONE ||)"
```

Comment:
Should obviously be fixed but when executing from the command line while it looks a bit odd to the 'parse error' the compiler refuses to compile (as it should) as well as it gives an indication the expression is incomplete at the right place.